### PR TITLE
feature: move test case runner. record unity test execution time into JUnit report

### PR DESCRIPTION
**Pull request by alekseiapa**
- Added unity test execution time into JUnit report
- Moved test_case_runner
- Added tests for test_case_runner

### CODE REVIEW

**Comment by hfudev** [comment]
@alekseiapa https://stackoverflow.com/questions/37471740/how-to-copy-commits-from-one-git-repo-to-another

**Comment by alekseiapa** [comment]
> @alekseiapa https://stackoverflow.com/questions/37471740/how-to-copy-commits-from-one-git-repo-to-another

@hfudev Yes, this is the way how I've done that. The author is @L-KAYA and the committer is me since I've changed the commit message to make it more appropriate for the current repository

**Comment by hfudev** [comment]
> > @alekseiapa https://stackoverflow.com/questions/37471740/how-to-copy-commits-from-one-git-repo-to-another
> 
> @hfudev Yes, this is the way how I've done that. The author is @L-KAYA and the committer is me since I've applied the commit to the current history and changed the commit message to make it more appropriate for the current repository. The commit history is totally preserved here

sorry! please ignore!

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (hfudev) [inline pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:118]: should also retry if this test case failed. here's the logic is a bit more complicated.

- if the result is passed:
    return success
- if the result is failed (or no result since core dumped or whatever) but retried:
    not decided yet
- if the result is failed (or no result since core dumped or whatever) and the retry_count has exceeded:
   return failed

Comment (hfudev) [reply pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:118]: this try...except... block should wrap the whole block, like

```python
try:
    self.dut.write(str(case.index))
    self.dut.expect(f'Running {case.name}...', timeout=1)
    self.dut.expect_unity_test_output(timeout=timeout)
except ...:
    ...
```

besides, the timeout check should also calculate the whole process., use some techniques like https://docs.python.org/3/library/threading.html#threading.Timer

Comment (hfudev) [reply pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:118]: same logic to other `run_...` functions

**Review by hfudev — CHANGES_REQUESTED** [review]
@alekseiapa Left a few comments about the retry logic.

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (L-KAYA) [inline pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:316]: As a noob, I learnt that this might be simplified to `if self.SEND_SIGNAL_PREFIX in match_str:`. Same as following `elif`

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (L-KAYA) [inline pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:20]: I forgot to delete this line, args is deleted

**Review by L-KAYA — COMMENTED** [review]
Thanks for moving this script into pytest-embeded!

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (alekseiapa) [reply pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py]: I suppose this logic can be applied to the `run_normal_case` only. For `run_multistage` and `run_multidev` it requires to pass the `sub_case_index`, so the `dut.write(str(sub_case_index))` can't be placed under `try` block

Comment (hfudev) [reply pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py]: what do you mean by "can't be placed under the try block" since those two functions "requires to pass the sub_case_index"?

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (alekseiapa) [reply pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:125]: @igrr Is your idea to only apply the retry logic at the start of the test ? But when we are checking the unity test output we shouldn't retry . Is that what you meant ?

**Comment by alekseiapa** [comment]
Hello @hfudev . Could you please take a look at the changes ? Thanks!

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (L-KAYA) [reply pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py]: When I tested the script in IDF, I found writing case index to dut might fail  after reset. I guess the failure mainly caused by the hurry writting before the dut get totally ready to receive something, because some delays will help. However, I can't set an exact delay since the required delay are different among different environment, especially on CI.  Therefore, I retired when failed to expect case running pattern.

As for the `expect_unity_test_output` I agree that it shouldn't be included in the `retry` block, because after expecting the `case running` pattern, the case is supposed to be running, when the timeout happened then, it should be treated as a failure.
cc @igrr @hfudev

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (L-KAYA) [reply pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:129]: > 



> this try...except... block should wrap the whole block, like
> 
> ```python
> try:
>     self.dut.write(str(case.index))
>     self.dut.expect(f'Running {case.name}...', timeout=1)
>     self.dut.expect_unity_test_output(timeout=timeout)
> except ...:
>     ...
> ```
> 
> besides, the timeout check should also calculate the spent time of the code block above. You may use https://docs.python.org/3/library/threading.html#threading.Timer or anything similar.

I would suggest to remain the legacy logic. Reasons given in https://github.com/espressif/pytest-embedded/pull/153#discussion_r1009061388

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (hfudev) [inline pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:22]: logging doesn't help. should raise a timeout and terminate running the current test case.

**Review by hfudev — CHANGES_REQUESTED** [review]

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (alekseiapa) [reply pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:22]: I agree with Raising timeout. My plan was to use this line later to track test execution time in the junit report.

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (hfudev) [reply pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py]: the original idea of the `retry` is for those flaky unit test cases. it should be parsed from the test case's full name. like `TEST_CASE(...[retry=3][timeout=120])`

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (hfudev) [reply pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:22]: this timeout would also help kill those hanging processes. that's why raise immediately is better, IMHO

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (hfudev) [reply pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:129]: I think we're talking about two things here... Let me clarify it a bit:

- for the idea to support keywords defined with `TEST_CASE` (described here: https://github.com/espressif/pytest-embedded/pull/153#discussion_r1009090953), please add the retry mechanism to the whole block
- for the idea to run retry while starting the test case, I agree with @L-KAYA 's idea. We could improve it later in a separate PR.

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (alekseiapa) [inline pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:19]: @hfudev  How about I apply the retry logic for the case in decorator: 
```python
def timeit(func):
    @wraps(func)
    def timeit_wrapper(*args, **kwargs):
        retry_count = int(kwargs.get('timeout', DEFAULT_RETRY_COUNT))
        start_time = time.perf_counter()
        for retry in range(retry_count):
            try:
                result = func(*args, **kwargs)
                break
            except TIMEOUT as e:
                if retry >= retry_count - 1:
                    raise e
        end_time = time.perf_counter()
        total_time = end_time - start_time
        logging.info(f'Testcase "{args[1].name}" took {total_time:.5f} seconds')

        if total_time > int(kwargs.get('timeout', DEFAULT_TIMEOUT)):
            raise TIMEOUT

        return result

    return timeit_wrapper
```

**Comment by alekseiapa** [comment]
@hfudev Could you please take a look at the changes ? Thank you!

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (hfudev) [inline pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:32]: > this timeout would also help kill those hanging processes. that's why raise immediately is better, IMHO

as I mentioned above, this current method can't kill the hanging processes.

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (hfudev) [inline pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:18]: - The `retry_count` here should be defined at `UnittestMenuCase.attributes['retry']`
- timeout should be defined at `UnittestMenuCase.attributes['timeout']`

> the original idea of the retry is for those flaky unit test cases. it should be parsed from the test case's full name. like TEST_CASE(...[retry=3][timeout=120])

As I mentioned above, these keywords should be defined in the `TEST_CASE(...)` c file. Not a keyword for python function.

**Review by hfudev — CHANGES_REQUESTED** [review]

**Comment by alekseiapa** [comment]
@hfudev I've added the logic of retrying the test case to the decorator. The options`UnittestMenuCase.attributes['retry']` and `UnittestMenuCase.attributes['timeout']` are parsed in `run_test` method

**Comment by alekseiapa** [comment]
@hfudev I've decided to resolve issue [#142](https://github.com/espressif/pytest-embedded/issues/142) in the current PR because it is related to the current task. What are your thoughts about it?

**Comment by hfudev** [comment]
Overall LGTM. btw, you don't need to mention the internal tickets here. I've removed them.

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (hfudev) [inline pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:52]: Instance attribute test_menu defined outside `__init__ `

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (hfudev) [inline pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:76]:
```suggestion
            reset: whether to do a hardware reset or not before running the case
```

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (hfudev) [inline pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:1]: rename this file to `unity_tester.py`, two `idf` is a bit redundant

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (hfudev) [inline pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:203]: could you help create a GH issue "replace threads in `MultiDevCaseTester`" later after your PR got merged?

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (hfudev) [inline pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:228]: this if clause seems a bit too defensive... consider remove it.

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (hfudev) [inline pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:29]:
```suggestion
    def __init__(self, dut: IdfDut, **kwargs) -> None:  # type: ignore
```

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (hfudev) [inline pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:196]:
```suggestion
    def __init__(self, dut: IdfDut) -> None:
```

**Review by hfudev — CHANGES_REQUESTED** [review]
a few more minor comments.

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (alekseiapa) [reply pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py:29]: I guess here it still has to be Union: 
`def __init__(self, dut: Union[IdfDut, List[IdfDut]], **kwargs) -> None:  # type: ignore`

Comment (hfudev) [reply pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py]: well, I think only `MultiDevTester` requires multiple duts?

let's keep it as what it is right now, and improve later.

File: pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py
Comment (hfudev) [reply pytest-embedded-idf/pytest_embedded_idf/idf_unity_tester.py]: >  after your PR got merged

anyway... I asked so because if this PR is not merged, the issue itself is meaningless.

**Comment by alekseiapa** [comment]
@hfudev Could you please take a look ? The pipeline is passing now

File: pytest-embedded-idf/pytest_embedded_idf/unity_tester.py
Comment (hfudev) [inline pytest-embedded-idf/pytest_embedded_idf/unity_tester.py:31]: IdfDut

**Review by hfudev — APPROVED** [review]

**Comment by hfudev** [comment]
thank you! @alekseiapa

**Pull request merged**
